Add field domain management capabilities to browser

nyalldawson opened the pull request:
This PR adds API to manage field domains to database connections, and exposes this functionality through the QGIS browser panel.

Some notes:

- This functionality is currently exposed only for Geopackages, but there is support in GDAL for field domain handling for other database types (ESRI FileGeodatabase, ESRI Personal Geodatabases) which could potentially be exposed in future. We could also potentially expose this functionality for the native database providers in future too (the API is designed to easily permit this)
- The functionality requires GDAL 3.5+
- QGIS already checks field domains when loading GPKG layers in order to setup nice defaults for editor widgets, but there's only partial support there (e.g. no support for glob domains in the QGIS editor widgets, or non-inclusive range domains). This functionality isn't changed by this PR.
- The new "Field Domains" item is only shown when expanding out GeoPackages which have some existing field domains (to avoid unwanted clutter)

TODO:

- [x] Expose ability to set the field domain for GPKG table fields
- [x] Expose ability to create new field domains
- [ ] Expose ability to delete and update field domain definitions (when supported by GDAL)

Demo showing display of existing domain properties:

![Peek 2022-01-27 15-14](https://user-images.githubusercontent.com/1829991/151296230-9776eede-b060-4b4f-980e-6bfde479271c.gif)

nyalldawson committed:
Create QgsFieldDomain and subclasses which directly represent
the GDAL field domain coded value/range/glob domains

nyalldawson committed:
Move logic for converting OGR field types to QVariant types to QgsOgrUtils

nyalldawson committed:
Add QgsOgrUtils::stringToVariant

nyalldawson committed:
Add method to convert OGR field domain to QgsFieldDomain equivalent

nyalldawson committed:
More repr

nyalldawson committed:
Add API to retrieve field domain names and definitions to QgsAbstractDatabaseProviderConnection
and implement for GPKG connections

nyalldawson committed:
Show a "Field Domains" item for geopackages under the browser,
which lists existing field domains in the database

nyalldawson committed:
Add python file

nyalldawson committed:
Add browser properties widget for showing the properties of an existing
field domain

nyalldawson commented:
@elpaso I'd love your eyes over this one!

nyalldawson committed:
Add docs

elpaso commented on src/app/browser/qgsinbuiltdataitemproviders.cpp:
```diff
@@ -1480,3 +1482,243 @@ void QgsDatabaseItemGuiProvider::populateContextMenu( QgsDataItem *item, QMenu *
+  metadata += QStringLiteral( "<tr><td class=\"highlight\">" ) + tr( "Type" ) + QStringLiteral( "</td><td>" ) + domain->typeName() + QStringLiteral( "</td></tr>\n" );
+
+  metadata += QStringLiteral( "<tr><td class=\"highlight\">" ) + tr( "Field type" ) + QStringLiteral( "</td><td>" );
+  switch ( domain->fieldType() )
```
It would be nice to have something in GUI utils to handles this: there is a risk of inconsistencies when the same type to string conversion is done in other parts of the code.

nyalldawson replied:
Ok, done. I've made a new QgsVariantUtils class which contains a static method for this (it's in core, since the providers and various models also need it)

elpaso replied:
@nyalldawson thanks!

elpaso reviewed (approved):
Nice addition!

nyalldawson committed:
Test requires GDAL 3.3+

nyalldawson committed:
Partial work on centralizing all QVariant::Type to display string
logic in a new QgsVariantUtils class

(I've likely missed a few places here, but it's a start!)

nyalldawson committed:
Split parts of QgsGeoPackageProviderConnection out to new base class
QgsOgrProviderConnection

This moves the generic parts of the GPKG connection class which rely
solely on OGR API out to a base class, leaving all the geopackage
specific functionality in the GPKG provider connection class.

It allows us to access a lot of the base level functionality from
this class for non-gpkg OGR supported database types, e.g.
spatialite, ESRI file geodatabases, MDB, etc.

nyalldawson committed:
Fix warnings

nyalldawson committed:
Update test

nyalldawson committed:
Fix setting capabilities for Gpkg connection

nyalldawson committed:
Silence debug

nyalldawson committed:
Add domain name to QgsFieldConstraints

nyalldawson committed:
[ogr] Set domain name for fields where available

nyalldawson committed:
Rename widget to generic name

nyalldawson committed:
[feature] Add browser properties page for field and fields items

In particular, this allows users to easily see what field domain
may be attached to particular fields

nyalldawson committed:
Add method to convert QVariant type to OGR field type

nyalldawson committed:
Add method to convert QVariant to OGRField

nyalldawson committed:
Add method to convert QgsFieldDomain to OGR field domain

nyalldawson committed:
Add API to QgsAbstractDatabaseProviderConnection to create new field
domains on a database and set field domain associated with a field

And implement for OGR database provider connection class

nyalldawson committed:
[feature] Show "Set Field Domain" action when right clicking
a field in the browser for a Geopackage table

Allows users to set the field domain associated with that field

nyalldawson commented:
It's now possible to view the details of which domain is attached to a field:

![Peek 2022-01-28 15-16](https://user-images.githubusercontent.com/1829991/151491287-4a4fc6d2-4955-478f-b3f9-047418cbc95b.gif)


And also set which domain is attached to a field:
![Peek 2022-01-28 15-17](https://user-images.githubusercontent.com/1829991/151491361-13c665ed-0eb2-4b3f-b968-6e702adedc57.gif)

nyalldawson committed:
Add flag for browser data item to indicate that children of the item
must be refreshed whenever the item is refreshed

This behaviour is not automatic. The new flag allows items to opt-in,
so that their children WILL be automatically refreshed when the
item is refreshed.

Should be used sparingly only to avoid expensive work

nyalldawson committed:
When a gpkg item is refreshed in the browser, also refresh
the fields and field domains children

Otherwise the fields we show will not be correctly updated
when the gpkg item itself is refreshed

nyalldawson committed:
QgsFieldItem is only equal when the QgsField is also equal

Otherwise the field item will not be updated when the gpkg parent
item is refreshed and the field definition has changed (e.g.
the field type)

nyalldawson committed:
Refresh fields item when a field has a domain set for it

nyalldawson committed:
Fix state of browser properties splitter section is not correctly
saved/restored

nyalldawson committed:
Clarify removal

nyalldawson committed:
Add action to unset an existing field domain

nyalldawson committed:
Fix documentation warning

nyalldawson committed:
Fix build on older gdal

nyalldawson committed:
Add widgets for creation of field domains

nyalldawson committed:
[feature] Allow creation of new field domains through right click
menu on gpkg items in browser

Allows users to create new range, coded value or glob field domains

nyalldawson committed:
Create QgsFieldDomainDialog dialog for configuring a field domain

nyalldawson committed:
Fix error retrieval

nyalldawson committed:
Minor tweaks

nyalldawson commented:
Last part is now implemented -- now it's possible to create new field domains through the browser for geopackages too!

![Peek 2022-02-16 10-42](https://user-images.githubusercontent.com/1829991/154174325-ffa11233-cd63-4464-88e5-0ba5756c0f60.gif)

nyalldawson committed:
Fix dox

nirvn commented:
OK, for the record, I'm _mega_ excited about this @nyalldawson ! :) Could we add a "load from CSV" functionality to the coded values domain addition dialog?

nyalldawson commented:
> Could we add a "load from CSV" functionality to the coded values domain addition dialog?

Sure, feel free to do that post merge :wink:

nirvn commented:
@nyalldawson , I shall -- I'll move what we have in the map value config dialog into QgsFileUtils or a new QgsCsvUtils class.

nyalldawson committed:
Fix test

agiudiceandrea commented:
Hi @nyalldawson, it seems this functionality has some issues and it could even lead to a crash in a circumstance. Please see https://github.com/qgis/QGIS/issues/52318.
